### ISSUE TITLE
LLDB: Support renaming symbols by address using a symbol file provided by JSON or other formatted data (e.g. protobuf, plist, xml, etc)

### ISSUE THREAD

**Issue by YungRaj** [llvm/llvm-project#99602]
Most reverse engineers would largely benefit from a feature that can allow for the symbolication of binaries to be achieved by setting or renaming the names of symbols via a LLDB python script. Oftentimes a linker map generated by a decompiler contains symbol names. Thus, if this feature was built, we can easily write an LLDB python script that parses the linker map file and populates the symbols in the binary by address.

**Comment by jimingham** [comment]
lldb already supports defining object files from JSON (ObjectFileJSON) including adding symbols.  Instead of using commands for defining symbols one-off, being able to create your symbol map in JSON and use extant machinery to get lldb to ingest it might be more useful.

**Comment by YungRaj** [comment]
Can you 

> lldb already supports defining object files from JSON (ObjectFileJSON) including adding symbols. Instead of using commands for defining symbols one-off, being able to create your symbol map in JSON and use extant machinery to get lldb to ingest it might be more useful.

Can you share with me a specific example on how this is achieved? StackOverflow answers say that it's not possible.

https://stackoverflow.com/questions/78354399/how-to-rename-a-stripped-unnamed-lldb-symbol

I haven't seen any specific projects that take advantage of this feature? Once I achieve this successfully, I can quickly go ahead and close the thread.

Thanks

**Comment by jimingham** [comment]
I wasn't trying to say "it works already" - I haven't tried to use ObjectFileJSON for exactly this usage.  Jonas added it to support the case where you have an artifact of a crash (a CrashLog on macOS) and while you can find some of the debug symbols locally, you can't find others, but you actually know SOME symbols from that library (they were in the CrashLog).  This allows you to add the symbols into the mix when parsing the CrashLog.
In that case, this is adding a NEW object file full of symbols, not overlaying extant one.  So you will likely have to do some lldb work to adapt the feature for this use case, but it shouldn't require all that much effort.

**Comment by YungRaj** [comment]
Well I guess my personal use case is not necessarily to symbolicate object files of libraries in e.g. userspace, but to symbolicate kernelcaches that are used in debugging the macOS/iOS kernel. Generally this matters more for iOS, but even when reverse engineering games in emulators etc this particular feature would be extremely useful.

And you're absolutely right about the linker map files being hard to parse and generally difficult to deal with, so I do agree to have LLDB to parse symbol names and addresses via JSON or even protobuf.

**Comment by YungRaj** [comment]
It looks like this feature was already implemented by https://github.com/llvm/llvm-project/commit/cf3524a5746f9498280b3a9180b75575c0065d1a

I'm gonna close this for now because it looks like it will work fine for my use cases.

For anyone with more questions, feel free to reopen this if you'd like.
